gax/src/main/java/com/google/api/gax/batching/BatcherImpl.java

```diff
@@ -201,6 +212,36 @@ private void awaitAllOutstandingBatches() throws InterruptedException {
     }
   }
 
+  /**
+   * It keeps the count of number of failed RPCs. This method also tracks the count for exception
+   * type along with counts for different failed {@link StatusCode}s.
+   */
+  private void addException(Throwable throwable) {
+    numOfFailure.incrementAndGet();
+    Class exceptionClass = throwable.getClass();
+
+    if (throwable instanceof ApiException) {
+      StatusCode code = ((ApiException) throwable).getStatusCode();
+      exceptionClass = ApiException.class;
+
+      synchronized (errorLock) {
+        if (failureStatusCodeCount.containsKey(code)) {
+          failureStatusCodeCount.get(code).incrementAndGet();
+        } else {
+          failureStatusCodeCount.put(code, new AtomicInteger(1));
+        }
+      }
+    }
+
+    synchronized (errorLock) {
+      if (failuresTypeCount.containsKey(exceptionClass)) {
+        failuresTypeCount.get(exceptionClass).incrementAndGet();
+      } else {
+        failuresTypeCount.put(exceptionClass, new AtomicInteger(1));
+      }
+    }
+  }
```

Comment:
@igorbernstein2 I have these two doubts regarding its implementation, Would you please have a look and let me know ur thoughts.
 1. In this change, I am adding the exception that occurred when RPC is being made, which means it would track exception thrown with `RequestT` rather than `ElementT`. 
 2. I hope using `Map<class, AtomicInteger>` to track the exception type should be fine.
 3. Having the count of `ApiException` is ok rather than having a count of each child of `ApiException`.

Reply:
1. I definitely think we need to track entry failure counts and I think it would be nice to track request level failures as well. Ideally we would report stats at both levels.
2. Yep thats fine
3. Yep counting the api in aggregate is good

I think useful messaging would look like
```
10 batches failed to apply: 5 NullPointerException, 2 ApiExceptions (1 DEADLINE_EXCEEDED, 1 INVALID_ARGUMENT), and 3 were partial failures. The 3 partial failures contained 10 entries that failed with DEADLINE_EXCEEDED, 20 entries with INVALID_ARGUMENT.
```

Reply:
Apologies for the delay, I have updated the message to similarily. Please have a look.